File: backend/dormammu/state/tasks.py
"""Parse operator-facing task documents (PLAN.md / TASKS.md) into structured state.

This module is responsible for turning the human-readable Markdown checklists
that dormammu writes to `.dev/PLAN.md` into machine-readable
:class:`OperatorTaskSyncState` objects consumed by the loop runner and daemon.

Recognised section titles (case-insensitive):
    - ``## Current Workflow``
    - ``## Prompt-Derived Development Queue``
    - ``## Prompt-Derived Implementation Plan``

Task lines inside those sections must match the format::

    - [ ] Pending task description
    - [O] Completed task description   (or [X] / [x])

A ``## Resume Checkpoint`` section, if present, is captured verbatim and
stored in :attr:`OperatorTaskSyncState.resume_checkpoint`.
"""
from __future__ import annotations

from dataclasses import dataclass
import re

from dormammu.state.models import OperatorTaskSyncState, TaskSyncItem


SECTION_HEADER_RE = re.compile(r"^##\s+(?P<title>.+?)\s*$")
TASK_LINE_RE = re.compile(r"^- \[(?P<marker>[ OXx])\] (?P<text>.+?)\s*$")


@dataclass(frozen=True, slots=True)
class ParsedTasksDocument:
    """The structured result of parsing a dormammu task document.

    Attributes:
        current_workflow: Parsed task items and optional resume checkpoint
            extracted from the active workflow section of the document.
    """

    current_workflow: OperatorTaskSyncState


def parse_tasks_document(text: str, *, source: str = ".dev/PLAN.md") -> ParsedTasksDocument:
    """Parse a dormammu PLAN.md or TASKS.md document.

    Args:
        text: Full Markdown content of the document to parse.
        source: Human-readable label used as the ``source`` field on the
            returned :class:`OperatorTaskSyncState` (defaults to
            ``".dev/PLAN.md"``).

    Returns:
        A :class:`ParsedTasksDocument` whose ``current_workflow`` attribute
        contains the discovered task items and optional resume checkpoint.
        Returns an empty workflow (no items, no checkpoint) when no recognised
        section is found.
    """
    queue_sections = {
        "current workflow",
        "prompt-derived development queue",
        "prompt-derived implementation plan",
    }
    current_section: str | None = None
    task_items: list[TaskSyncItem] = []
    resume_lines: list[str] = []

    for raw_line in text.splitlines():
        header = SECTION_HEADER_RE.match(raw_line)
        if header:
            current_section = header.group("title").strip().lower()
            continue

        if current_section in queue_sections:
            task_match = TASK_LINE_RE.match(raw_line)
            if task_match:
                marker = task_match.group("marker")
                task_items.append(
                    TaskSyncItem(
                        text=task_match.group("text").strip(),
                        completed=marker.upper() in {"O", "X"},
                    )
                )
        elif current_section == "resume checkpoint":
            stripped = raw_line.strip()
            if stripped:
                resume_lines.append(stripped)

    return ParsedTasksDocument(
        current_workflow=OperatorTaskSyncState(
            source=source,
            resume_checkpoint="\n".join(resume_lines) or None,
            items=tuple(task_items),
        )
    )

File: runtime/src/state/tasks.ts
const SECTION_HEADER_RE = /^##\s+(?<title>.+?)\s*$/;
const TASK_LINE_RE = /^- \[(?<marker>[ OXx])\] (?<text>.+?)\s*$/;

const QUEUE_SECTIONS = new Set([
  "current workflow",
  "prompt-derived development queue",
  "prompt-derived implementation plan"
]);

export type TaskSyncItem = {
  text: string;
  completed: boolean;
};

export type OperatorTaskSyncState = {
  source: string;
  resumeCheckpoint: string | null;
  items: readonly TaskSyncItem[];
};

export type OperatorTaskSyncPayload = {
  source: string;
  resume_checkpoint: string | null;
  total_tasks: number;
  completed_tasks: number;
  pending_tasks: number;
  all_completed: boolean;
  next_pending_task: string | null;
  items: TaskSyncItem[];
  synced_at: string;
};

export type ParsedTasksDocument = {
  currentWorkflow: OperatorTaskSyncState;
};

export function operatorTaskSyncToDict(
  state: OperatorTaskSyncState,
  options: { syncedAt: string }
): OperatorTaskSyncPayload {
  const totalTasks = state.items.length;
  const completedTasks = state.items.filter((item) => item.completed).length;
  const pendingTasks = totalTasks - completedTasks;
  const nextPendingTask = state.items.find((item) => !item.completed)?.text ?? null;
  return {
    source: state.source,
    resume_checkpoint: state.resumeCheckpoint,
    total_tasks: totalTasks,
    completed_tasks: completedTasks,
    pending_tasks: pendingTasks,
    all_completed: totalTasks > 0 && pendingTasks === 0,
    next_pending_task: nextPendingTask,
    items: state.items.map((item) => ({ text: item.text, completed: item.completed })),
    synced_at: options.syncedAt
  };
}

export function parseTasksDocument(
  text: string,
  options: { source?: string } = {}
): ParsedTasksDocument {
  let currentSection: string | null = null;
  const items: TaskSyncItem[] = [];
  const resumeLines: string[] = [];

  for (const rawLine of text.split(/\r?\n/)) {
    const header = SECTION_HEADER_RE.exec(rawLine);
    if (header?.groups?.title) {
      currentSection = header.groups.title.trim().toLocaleLowerCase();
      continue;
    }

    if (currentSection !== null && QUEUE_SECTIONS.has(currentSection)) {
      const taskMatch = TASK_LINE_RE.exec(rawLine);
      if (taskMatch?.groups?.marker && taskMatch.groups.text) {
        const marker = taskMatch.groups.marker;
        items.push({
          text: taskMatch.groups.text.trim(),
          completed: marker.toLocaleUpperCase() === "O" || marker.toLocaleUpperCase() === "X"
        });
      }
    } else if (currentSection === "resume checkpoint") {
      const stripped = rawLine.trim();
      if (stripped) {
        resumeLines.push(stripped);
      }
    }
  }

  return {
    currentWorkflow: {
      source: options.source ?? ".dev/PLAN.md",
      resumeCheckpoint: resumeLines.length ? resumeLines.join("\n") : null,
      items
    }
  };
}
